src/libsyntax/print/pprust.rs

```diff
@@ -1651,9 +1651,17 @@ impl<'a> State<'a> {
                 }
             }
             ast::StmtKind::Semi(ref expr) => {
-                self.space_if_not_bol();
-                self.print_expr_outer_attr_style(expr, false);
-                self.s.word(";");
+                match expr.node {
+                    // Filter out empty Tup exprs created for the redundant_semicolon
+                    // lint, as they shouldn't be visible and interact poorly
+                    // with proc macros
+                    ast::ExprKind::Tup(ref exprs) if exprs.is_empty() => (),
```

Comment:
We can use a slice pattern instead to express this more succinctly:
```suggestion
                    ast::ExprKind::Tup([]) => {}
```